x/slashing/internal/keeper/infractions.go

```diff
@@ -184,6 +218,44 @@ func (k Keeper) HandleValidatorSignature(ctx sdk.Context, addr crypto.Address, p
 			// That's fine since this is just used to filter unbonding delegations & redelegations.
 			distributionHeight := height - sdk.ValidatorUpdateDelay - 1
 
+			// Insert slash event into the recent liveness faults iterator
+			k.InsertLivenessQueue(ctx, types.NewSlashEvent(
+				validator.GetConsAddr(),
+				power,
+				distributionHeight,
+				sdk.NewDec(power).QuoInt(k.sk.GetLastTotalPower(ctx)),
+				sdk.ZeroDec(),
+				ctx.BlockHeader().Time.Add(k.DowntimeJailDuration(ctx)),
+			))
+
+			// iterate over all recent slashes to get the sum of the roots of all the powers
+			squareOfSumOfRoots := sdk.ZeroDec()
+			k.IterateLivenessQueue(ctx, func(slashEvent types.SlashEvent) bool {
+				squareOfSumOfRoots = squareOfSumOfRoots.Add(slashEvent.PercentPower.RoughSqrt())
+				return false
+			})
+			// square the sum of the roots
+			squareOfSumOfRoots = squareOfSumOfRoots
+			// get the percentage slash penalty fraction by multiplying by constant scalar from param store
+			fraction := k.SlashFractionDoubleSign(ctx).Mul(squareOfSumOfRoots)
+
+			k.IterateLivenessQueue(ctx, func(slashEvent types.SlashEvent) bool {
+				// If a validator needs to be further slashed
+				if fraction.GT(slashEvent.SlashedSoFar) {
+					// how much additional to slash the validator
+					leftToSlash := fraction.Sub(slashEvent.SlashedSoFar)
+
+					// slash the validator at leftToSlash amount
+					k.sk.Slash(ctx, slashEvent.Address, slashEvent.InfractionHeight, slashEvent.Power, leftToSlash)
+
+					// update slash so far amount in slash event
+					slashEvent.SlashedSoFar = fraction
+					k.InsertDoubleSignQueue(ctx, slashEvent)
```

Comment:
Why are we inserting into the double-sign queue? Also, when is this queue cleared?

Reply:
Sorry this PR was meant to be a WIP, not a R4R.  But just added the clearing of the queue in the most recent commit.

Adding it into the double-sign queue, because it can be processed the same way as everything in the queue already as long as the `SlashedSoFar` is set to 0.  Then it will just be slashed at the `SquareOfSumOfRoots` amount.